src/lib/support/PersistentArray.h

```diff
@@ -0,0 +1,316 @@
+
+        // Entries
+        ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(PersistentTags::kList), TLV::kTLVType_List, list));
+        for (size_t i = 0; (i < this->mCount) && (i < this->mLimit); ++i)
+        {
+            ReturnErrorOnFailure(Serialize(writer, mEntries[i]));
+        }
+        ReturnErrorOnFailure(writer.EndContainer(list));
+
+        return writer.EndContainer(container);
+    }
+
+    CHIP_ERROR Deserialize(TLV::TLVReader & reader) override
+    {
+        ReturnErrorOnFailure(reader.Next(TLV::AnonymousTag()));
+        VerifyOrReturnError(TLV::kTLVType_Structure == reader.GetType(), CHIP_ERROR_INTERNAL);
+
+        TLV::TLVType container, list;
+        ReturnErrorOnFailure(reader.EnterContainer(container));
+
+        // Count
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(PersistentTags::kCount)));
+        ReturnErrorOnFailure(reader.Get(this->mCount));
+
+        // Entries
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(PersistentTags::kList)));
+        VerifyOrReturnError(TLV::kTLVType_List == reader.GetType(), CHIP_ERROR_INTERNAL);
+
+        ReturnErrorOnFailure(reader.EnterContainer(list));
+        for (size_t i = 0; i < this->mCount && i < this->mLimit; ++i)
+        {
+            ReturnErrorOnFailure(Deserialize(reader, mEntries[i]));
+        }
+        ReturnErrorOnFailure(reader.ExitContainer(list));
+
+        return reader.ExitContainer(container);
+    }
+
+    /**
+     * Stores an entry at a given position. This method does not validates the order
+     * in which the entries are stored. New entries must be inserted with index == count,
+     * if index < count, the existing entry is replaced, generating an implicit remove of
+     * the existing entry.
+     * The Compare() method is used to ensure that the entry is unique. If the new entry matches
+     * an existing entry, the index must also match, otherwise CHIP_ERROR_DUPLICATE_KEY_ID is
+     * returned.
+     * @param index Zero-based position within the table.
+     * @param entry Value to be store in the table.
+     * @return CHIP_NO_ERROR on success. CHIP_ERROR_DUPLICATE_KEY_ID if the entry is already
+     * stored in different index. CHIP_ERROR_INVALID_ARGUMENT if index >= limit.
+     */
+    CHIP_ERROR Set(size_t index, const EntryType & value)
+    {
+        CHIP_ERROR err = this->Load();
+        VerifyOrReturnError(CHIP_ERROR_NOT_FOUND == err || CHIP_NO_ERROR == err, err);
+
+        // Check existing
+        for (size_t i = 0; i < this->mCount; ++i)
+        {
+            EntryType & e = mEntries[i];
+            if (this->Compare(e, value))
+            {
+                // Existing value, index must match
+                VerifyOrReturnError(i == index, CHIP_ERROR_DUPLICATE_KEY_ID);
+                ReturnErrorOnFailure(this->Copy(e, value));
+                return this->Save();
```

Comment:
So this case does not call OnEntryRemoved or OnEntryAdded, right?

Why do we have those hooks?  Is anything using them?